### PR TITLE
Spec for Nodes

### PR DESCRIPTION
**Pull request by mpu-creare**
This is a spec describing the basic design of Nodes.

### CODE REVIEW

**Comment by mpu-creare** [comment]
Had a chat with @jmilloy about `params`. We talked through a number of use-cases and decided we can eliminate the 'params' call from the eval call
* It's hard to decide what's a `param` and what's an `attr`. Our users will have the same problem
* Default params and default attrs behave the same
* Overwriting a param at evaluation on the Python side has limited uses -- you can always just change the parameter directly in the pipline
* Overwriting a param at evaluation on the JSON side has no use case -- you always create the pipeline through the JSON to make a request. So essentially the defaults that are specified are used. 
* If you need the SAME node with TWO different params, just create TWO instances of the node with the different params
    * Nodes are pretty lightweight as it stands
    * With the new caching system there is no performance hit for retrieving data -- both nodes will look at the same cache. 
* Eliminating the params simplifies the complex passing of parameters to sub-nodes in  pipelines
    * Previously you needed to specify params={node1.node2.node3.param: value, ...} to set the correct param on the correct node

**Comment by jmilloy** [comment]
See #87 for the issue with this change specifically.

@mpu-creare I will implement this week in this branch unless you have an objection.

**Comment by jmilloy** [comment]
I believe that this issue is also part of this PR #88

**Comment by mpu-creare** [comment]
Agreed. That works for me. I'll make some adjustments to `Node` on top of these changes on this branch as well. How do you want to stage the merges? Pipeline, Coords, Nodes?

**Comment by jmilloy** [comment]
Actually, I'd like to do the params->attrs in its own branch, either after
the pipeline merge if you can do that soon or at least based off the
feature/pipeline-node branch so I can do it this week. Then Coords. Then
what's left for Nodes (node evaluation checks, ...?).

**Comment by mpu-creare** [comment]
I don't have a lot of bandwidth right now, so I won't be able to do anything soon. So please do whatever suits you.

**Comment by jmilloy** [comment]
@mpu-creare looks great!

 * I added some info about missing/extra *stacked* dimensions, please take a look at that. We can drop extra unstacked dimensions because they are actually xarray dimensions and it will still broadcast correctly, but I think we need to keep extra stacked coordinates. And for missing unstacked dimensions, we throw an exception, but for missing stacked dimensions, we can just drop them as long as the resulting coordinates are still unique, as we discussed.
 * I propose removing the most recent `evaluated_coordinates` from the public interface, similar to removing the `output` from the public interface. Same reasoning.
 * I also did some housekeeping:
    * removed the evaluation params from the spec
    * updated execute to eval/evaluate
 * We decided to disallow evaluation of GroupCoordinates in favor of an `eval_group` helper function, but I didn't update that in the spec.
 * I haven't thought about the units much, do you want me to take a longer look at that?

**Comment by jmilloy** [comment]
@mpu-creare One thing that I find difficult with the Node class, as a developer, is the various initialize_array methods.

One thing is that you have to remember to set `self.evaluated_coordinates` in the `eval` method before you can call `get_output_coords`, `get_output_shape`, `initialize_output_array`, etc without a coords argument. I think these methods should require the coords explicitly. I further propose that the initialize methods be removed from the Node class entirely and into a util module as standalone functions. Any functions that remain in the Node class and use evaluated_coordinates needs to be private.

For the initialize methods themselves, I always have to go look at their code in node.py to remember which one does what and takes which arguments. I think requiring explicit arguments will also simplify their implementation, make reading code that uses these functions clearer, and allow us to remove some of them and/or give them clearer names. For example, they in fact don't initialize anything.

Do we really need copy_output_array? Isn't it just two lines?

```
x = self._output.copy(True)
x.data[:] = np.nan
```

**Comment by mpu-creare** [comment]
I agree with all your comments. I also always have to look up what the initialize methods do. The one non-obvious thing is that they populate metadata (like styling information) from the node. But I suppose a utility function could do that as well. Perhaps the constructor to UnitsDataArray could handle all of this...

**Comment by jmilloy** [comment]
* The `native_coordinates` will only be defined in DataSource nodes. Likewise, we will only need to check the requested coordinates against the native coordinates in the DataSource eval.
 * The `interpolation` will be defined in both DataSource and Compositor nodes, but not Algorithm or Pipeline nodes.
 * The `Node` will have matching `get_cache`, `put_cache`, and `del_cache` methods, e.g. `get_cache(self, 'output', coordinates)`
 * Nodes should not be re-used within the same pipeline; one big benefit of this is that it simplifies debugging in cases where the evaluated coordinates change -- we won't need `__get_item__` as specified above.

**Comment by drc-creare** [comment]
For the `Node` cache methods (`get_cache`, `put_cache`, and `del_cache`), does there need to be a mechanism for specifying that the cached data is either 1) shared among any `Node` object that has the same pipeline definition or 2) is specific to the exact `Node` object?  Or do we not need the latter? I'm thinking about when we are using cache to debug pipeline/Node execution.

**Comment by mpu-creare** [comment]
> Do you ever want to cache objects that are independent of the coordinates (such as native_coordinates) through the public interface put_cache

Yes. We want people who write their own nodes to be able to cache things that they care about. Things they care about might be independent of coordinates. 

> should the coordinates argument be optional for get_cache and put_cache
Yes.

**Comment by jmilloy** [comment]
@drc-creare For case 2, I guess you could just put the object id into the key if that's what you needed:

```
key = 'my_key_%s' % id(self)
self.put_cache(key, coordinates, data)
cached_data = self.get_cache(key, coordinates)
```

**Comment by drc-creare** [comment]
I vote for `coordinates` being an optional argument for Node's public interface to the cache. We could probably handle `native_coordinates` caching automatically under-the-hood. But, I think there could be other things that would be coordinate agnostic. I think SMAP save "shared_coordinates" for instance and I dont think that this is a property inherited from Node.

**Comment by drc-creare** [comment]
@jmilloy Yeah, that use of `id()` works. Also, just spoke to @mpu-creare and we couldn't come up with a reason for the unique object cache. So we should probably not worry about until/if we need it.

**Comment by jmilloy** [comment]
@mpu-creare Thanks, spec updated again.

**Comment by jmilloy** [comment]
This is now only waiting on #124.

File: podpac/core/data/types.py [hunk 1]
```diff
@@ -247,9 +246,8 @@ def get_native_coordinates(self):
     def get_data(self, coordinates, coordinates_index):
         """{get_data}
         """
-        data = self.dataset[self.datakey][tuple(coordinates_index)]
-        d = self.initialize_coord_array(coordinates, 'data',
-                                        fillval=data.reshape(coordinates.shape))
+        data = self.dataset[self.datakey][tuple(coordinates_index)].data
```
[the remark this answers is not in the document]
Comment (jmilloy) [reply podpac/core/data/types.py]: For me, the unit pydap get_data test fails without the `.data`. My pydap is version 3.2.2.

File: podpac/core/node.py [hunk 2]
```diff
@@ -220,241 +159,82 @@ def init(self):
         pass
 
     @common_doc(COMMON_DOC)
-    def execute(self, coordinates, output=None, method=None):
-        """This is the common interface used for ALL nodes. Pipelines only
-        understand this method and get_description.
+    def eval(self, coordinates, output=None, method=None):
```
[the remark this answers is not in the document]
Comment (jmilloy) [reply podpac/core/node.py]: 1e21dad

File: podpac/core/node.py [hunk 3]
```diff
@@ -491,51 +277,39 @@ def base_definition(self):
...
-        See Also
-        --------
-        base_definition
+        OrderedDict
+            Dictionary-formatted definition of a PODPAC pipeline. 
         """
 
-        raise NotImplementedError
+        from podpac.core.pipeline import make_pipeline_definition
+        return make_pipeline_definition(self)
```
[the remark this answers is not in the document]
Comment (jmilloy) [reply podpac/core/node.py]: Nope, I think that's no longer necessary.

b61bbb4

File: podpac/core/data/types.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mpu-creare) [reply podpac/core/data/types.py]: Yep. I have 3.2.1, and the `.data` makes IT fail. Great, so they changed something. I'll make an update to handle both version.

**Comment by mpu-creare** [comment]
I originally had the `def algorithm(self, **inputs)` (very first version of Code), and feedback was that it was confusing. I think needing to *know* that you have to use self.outputs is also not great. I think what you have now with the `inputs` is a perfect middle-ground.

**Comment by mpu-creare** [comment]
I'm happy with this. I'll merge in once the check finish.

**Pull request merged**